File: src/lib/ipc/picus/binds.ts
/**
 * Running a statement with its values bound.
 *
 * A door of its own rather than an argument on {@link execute}, because the two
 * differ in what they hand back: a cursor cannot take parameters, so a bound read
 * opens **no scrollable result** — `resultId` is null and `endOfResult` says
 * whether anything was left behind. Everything else about the reply is the same
 * `ExecuteResult` the rest of the studio already reads.
 *
 * The values never enter the SQL text. That is the entire point of the feature and
 * it is enforced in the driver, not here: the statement crosses the wire with its
 * `$1` intact and the value travels in the protocol's own field, where nothing can
 * read it as syntax.
 */

import { picus } from '../rpc';
import type { ExecuteResult } from './db';

/**
 * One value on the wire.
 *
 * `null` is a real SQL NULL and **not** the empty string — the two are different
 * rows on a text column, and the modal asks for them separately for that reason.
 * Picus sends strings: what the user typed, parsed by the server with the input
 * function of the type it inferred for that placeholder, which is what makes its
 * complaint about a bad value the server's own words.
 */
export type BindValue = string | number | boolean | null;

/**
 * Run one statement with `binds` bound to its placeholders.
 *
 * `binds` is **positional**: entry 0 is `$1` on PostgreSQL, and the first
 * placeholder in order of appearance on Oracle. Build it with `toBindList` from
 * `sql-intel/binds`, which knows both orders.
 *
 * Only for a connection whose engine reports `capabilities.bindParameters` — an
 * engine without it must not be offered the flow at all, rather than offered it and
 * refused.
 */
export function executeBound(
  connectionId: string,
  sql: string,
  binds: BindValue[],
  window?: number,
): Promise<ExecuteResult> {
  return picus('picus_execute_bound', { connectionId, sql, binds, window });
}

File: crates/products/picus/be/src/binds.rs
//! `binds` domain — running a statement whose values are sent beside it.
//!
//! The sibling of [`picus_execute`](crate::query::picus_execute), and deliberately
//! a **second door** rather than an optional argument on the first one. Two reasons,
//! and both are about what a caller can be sure of:
//!
//! * binding is a **capability**. `EngineCapabilities::bind_parameters` says whether
//!   an engine has it, and the interface reads the flag instead of calling and
//!   catching — a feature the engine lacks has to be absent, not present and
//!   failing. A `binds` key quietly ignored by an engine that cannot bind would be
//!   exactly the second kind;
//! * the **result differs**, and it differs in a way the caller must see. A cursor
//!   is a utility statement and takes no parameters, so a bound read opens no
//!   scrollable result: `resultId` is null and `endOfResult` is the honest statement
//!   of whether anything was left behind. Folding that into `picus_execute` would
//!   make the same call sometimes scrollable and sometimes not.
//!
//! The wire keys are `connectionId`, so the argument is spelled that way — the
//! contract wins over the naming convention, as it does in [`crate::query`].
#![allow(non_snake_case)]

use picus_core::prelude::PicusState;
use picus_db_api::prelude::{BindValue, ExecuteResult};

use crate::connections::require_session;
use crate::query::window_size;

/// Run a statement with its values bound to its placeholders.
///
/// `binds` is **positional**: `binds[0]` is `$1` on PostgreSQL and the first
/// placeholder in order of appearance on Oracle. A `null` entry is a real SQL NULL
/// and not the empty string — the interface asks for the two separately, because in
/// a maintenance tool confusing them is how a bad `UPDATE` gets written.
///
/// Nothing here inspects the SQL: what a statement is, how many values it wants and
/// whether it produces rows are all questions the engine asks the server. A caller
/// that had to classify SQL first would get it wrong silently.
#[arbor_rpc::handler]
async fn picus_execute_bound(
    state: &PicusState,
    connectionId: String,
    sql: String,
    binds: Vec<BindValue>,
    window: Option<u32>,
) -> Result<ExecuteResult, String> {
    require_session(state, &connectionId)?
        .execute_bound(&sql, &binds, window_size(window))
        .await
        .map_err(|e| e.to_string())
}
